src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/ExtractToCodeBehindCodeActionProvider.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.CodeAnalysis;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Refactoring
+{
+    class ExtractToCodeBehindCodeActionProvider : RazorCodeActionProvider
+    {
+        override public Task<CommandOrCodeActionContainer> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            var directiveNode = (RazorDirectiveSyntax)context.Document.GetNodeAtLocation(context.Location, n => n.Kind == SyntaxKind.RazorDirective);
+            if (directiveNode is null)
+            {
+                return null;
+            }
+
+            if (directiveNode.DirectiveDescriptor.Directive != "code" && directiveNode.DirectiveDescriptor.Directive != "function")
```

Comment:
The directive descriptors should be compared against the actual directives: 

- `@code` https://github.com/dotnet/aspnetcore/blob/9e5ee0665671f45debe204d75c4100c4f7b007b0/src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentCodeDirective.cs#L10
- `@functions` https://github.com/dotnet/aspnetcore/blob/9e5ee0665671f45debe204d75c4100c4f7b007b0/src/Razor/Microsoft.AspNetCore.Razor.Language/src/Extensions/FunctionsDirective.cs#L11